xen-exporter.py

```diff
@@ -139,109 +139,120 @@ def collect_metrics():
+                        extra_tags["sr_uuid"] = long_sr
+                    metric_type = "_".join(metric_type.split("_")[0:-1])
+
+                if collector_type == "vm" and "vbd_" in metric_type:
+                    x = metric_type.split("vbd_")[1]
+                    extra_tags["vbd"] = x.split("_")[0]
+                    metric_type = "vbd_" + "_".join(x.split("_")[1:])
+
+                if collector_type == "vm" and "vif_" in metric_type:
+                    x = metric_type.split("vif_")[1]
+                    extra_tags["vif"] = x.split("_")[0]
+                    metric_type = "vif_" + "_".join(x.split("_")[1:])
+
+                if collector_type == "host" and "pif_" in metric_type:
+                    x = metric_type.split("pif_")[1]
+                    extra_tags["pif"] = x.split("_")[0]
+                    metric_type = "pif_" + "_".join(x.split("_")[1:])
+
+                if "cpu" in metric_type:
+                    x = metric_type.split("cpu")[1]
+                    if x.isnumeric():
+                        extra_tags["cpu"] = x
+                        metric_type = "cpu"
+                    elif "-" in x:
+                        extra_tags["cpu"] = x.split("-")[0]
+                        metric_type = "cpu_" + x.split("-")[1]
+                if "CPU" in metric_type:
+                    x = metric_type.split("CPU")[1]
                     extra_tags["cpu"] = x.split("-")[0]
-                    metric_type = "cpu_" + x.split("-")[1]
-            if "CPU" in metric_type:
-                x = metric_type.split("CPU")[1]
-                extra_tags["cpu"] = x.split("-")[0]
-                metric_type = "cpu_" + "_".join(x.split("-")[1:])
-
-            # Normalize metric names to lowercase and underscores
-            metric_type = metric_type.lower().replace("-", "_")
-
-            tags = {f'{k}="{v}"' for k, v in extra_tags.items()}
-            output += f"xen_{collector_type}_{metric_type}{{{', '.join(tags)}}} {metrics['data'][0]['values'][i]}\n"
-
-        output += collect_sr_usage(xen)
-        collector_end_time = time.perf_counter()
-        output += f"xen_collector_duration_seconds {collector_end_time - collector_start_time}\n"
-        return output
+                    metric_type = "cpu_" + "_".join(x.split("-")[1:])
+
+                # Normalize metric names to lowercase and underscores
+                metric_type = metric_type.lower().replace("-", "_")
+
+                tags = {f'{k}="{v}"' for k, v in extra_tags.items()}
+                output += f"xen_{collector_type}_{metric_type}{{{', '.join(tags)}}} {metrics['data'][0]['values'][i]}\n"
+            output += collect_sr_usage(xen)
+    except BaseException as e:
+        print(e)
+        xen_host_up = False
+    collector_end_time = time.perf_counter()
+    output += collect_sr_usage(xen)
+    output += f"xen_collector_duration_seconds {collector_end_time - collector_start_time}\n"
+    xen_host_up_extra_tags = {}
```

Comment:
This block seems quite a bit more complicated than necessary. You just need to output the xen_host_up value and single ip tag. In this case the extra tags dict always only has one value in it.